src/shogun/optimization/liblinear/shogun_liblinear.cpp

```diff
@@ -43,9 +43,52 @@
 #include <shogun/optimization/liblinear/tron.h>
 #include <shogun/lib/Time.h>
 #include <shogun/lib/Signal.h>
+#include <omp.h>
 
 using namespace shogun;
 
+Reduce_Vectors::Reduce_Vectors(int size)
+{
+	nr_thread = omp_get_max_threads();
+	this->size = size;
+	tmp_array = new double*[nr_thread];
+	for(int i = 0; i < nr_thread; i++)
+		tmp_array[i] = new double[size];
+}
+
+Reduce_Vectors::~Reduce_Vectors(void)
+{
+	for(int i = 0; i < nr_thread; i++)
+		delete[] tmp_array[i];
+	delete[] tmp_array;
+}
+
+void Reduce_Vectors::init(void)
+{
+#pragma omp parallel for schedule(static)
+	for(int i = 0; i < size; i++)
+		for(int j = 0; j < nr_thread; j++)
+			tmp_array[j][i] = 0.0;
+}
+
+void Reduce_Vectors::sum_scale_x(double scalar, CDotFeatures *x, int i, int n)
+{
+	int thread_id = omp_get_thread_num();
+
+	x->add_to_dense_vec(scalar, i, tmp_array[thread_id], n);
+}
+
+void Reduce_Vectors::reduce_sum(double* v, int n)
+{
+#pragma omp parallel for schedule(static)
+	for(int i = 0; i < n; i++)
+	{
```

Comment:
Yes, I was curious about this too, how eigen3 likes would fare here. I have kept the authors way of doing things for now.

yes this could be added in some way to linalg. Could be useful if one wants to reduce arrays/vectors after working with them in multiple threads.